File: artifacts/tulipindicators/translate_new_ablation/src/roc.rs
use crate::*;
use rand::Rng;

pub fn ti_roc_start(options: Option<&[f64]>) -> i32 {
    match options {
        Some(opts) if !opts.is_empty() => opts[0] as i32,
        _ => 0, // Return 0 for NULL pointer or empty slice
    }
}

pub fn ti_roc(size: i32, inputs: &[&[f64]], options: &[f64], outputs: &mut [&mut [f64]]) -> i32 {
    let input = inputs[0];
    let period = options[0] as i32;
    let output = &mut outputs[0];
    
    if period < 1 {
        return 1;
    }
    
    let start = ti_roc_start(Some(options));
    if size <= start {
        return 0;
    }
    
    let mut output_idx = 0;
    for i in period..size {
        let value = (input[i as usize] - input[(i - period) as usize]) / input[(i - period) as usize];
        output[output_idx] = value;
        output_idx += 1;
    }
    
    assert!((output_idx as i32) == (size - ti_roc_start(Some(options))));
    0
}

File: artifacts/tulipindicators/decompose/src/roc.c
int ti_roc_start(const double *options)
{
  return (int) options[0];
}


/*** DEPENDENCIES:
***/


int ti_roc(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_roc_start(options))
  {
    return 0;
  }
  int i;
  for (i = period; i < size; i += 1)
  {
    output[output_idx] = (input[i] - input[i - period]) / input[i - period];
    output_idx += 1;
  }

  assert(((&output[output_idx]) - outputs[0]) == (size - ti_roc_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_roc_start(const double *options)
{
  return (int) options[0];
}


----------------------------
***/


